doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+compute cluster also means,
+- We eventually will support `SELECT` queries in a `COPY ... TO S3` command and that can
+be a natural extension if things are already in compute.
+- Via the Persist Compute Sink, there's already precedence to talking to S3, we should
+be able to leverage some existing code.
+- This also will be consistent with the behaviour that a user can select the compute cluster
+where they want their query to run like they do for other queries.
+
+#### Other Alternative: Storage cluster
+The other alternative would have been to actually make use of a Storage cluster. If we think
+of `COPY ... TO S3` as a one time sink, then it might also make sense to have it in the Storage
+cluster. Especially if the initial scope does not include `SELECT` queries which
+would have required spinning up compute dataflows.
+Reasons to not do so,
+- As mentioned above, we will eventually support `SELECT` queries in `COPY ... TO S3`. The
+only reason it would have made sense to do this in storage would be to do as a quick stop gap
+if the effort is drastically less. It does not appear to be so.
+- Unlike Compute, Storage does not have a concept of running temporary/one-off dataflows.
+Everything which runs, like sources or sinks, do so continuously once scheduled till they are
+dropped.
+
+### Output file format
+We should start with a single file format and add more if there's more user interest. Currently,
+there has been request for parquet but CSV support would be easiest to build.
+
+Postgres supports `COPY ... TO` with a csv file and we should mimic the file structure for our
+export as well.
+
+Eventually we'll need to add some additional formatting options to the SQL like
+`DELIMITER` (default will be comma), `QUOTE` (default is double quotes) etc. In the initial
+version we should stick to the default values as per the
+[Postgres COPY command](https://www.postgresql.org/docs/current/sql-copy.html).
+
+### Bounded Memory
+I had wrongly assumed that with reading directly from persist we could use backpressure similar to what we do in upsert. As @guswynn correctly pointed out, that wouldn't work due to the following:
+* For upsert, we read configured amount of data from persist, but we eventually consolidate in rocksdb (or in memory if disk is not enabled) which maintains the upsert state.
+* In compute, we may be able to fetch data in chunks from persist, but it will still need to be
+consolidated for the entire shard and this will have to happen in memory.
+
+One possibility could be that we actually use disk (with rocksdb) in Compute to keep state similar
+to upsert and consolidate the data there before writing the snapshot to S3. This seems complicated
+and I don't think it will be worth the effort. Also, given that we might
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+TODO: Check on the status of consolidation on read from persist, if it can be used here.
```

Comment:
> That does mean we should be careful with buffers and such in the implementation of COPY TO S3—we don't want the implementation to introduce any new uses of memory.

Cool, that leaves arrangements as the only option because we still need to do consolidation and arrangements is the only way to not introduce new uses of memory.

Reply:
Oh, sorry, I think it's fine to stick a `.consolidate()` operator where you need it, while `persist_source` doesn't do consolidation on read. I just want to make sure that once persist supports CoR, it's ~no work to make `COPY persist_backed_relation TO S3` use bounded memory.

Reply:
Updated section to mention that we should be careful with memory, and we are able to leverage persist consolidate on read when available.